src/scores/probability/roc_impl.py

```diff
@@ -0,0 +1,126 @@
+        An xarray.Dataset with data variables:
+
+        - 'POD' (the probability of detection)
+        - 'POFD' (the probability of false detection)
+        - 'AUC' (the area under the ROC curve)
+
+        `POD` and `POFD` have dimensions `dims` + 'threshold', while `AUC` has
+        dimensions `dims`.
+
+    Notes:
+        The probabilistic `fcst` is converted to a deterministic forecast
+        for each threshold in `thresholds`. If a value in `fcst` is greater
+        than or equal to the threshold, then it is converted into a
+        'forecast event' (fcst = 1), and a 'forecast non-event' (fcst = 0)
+        otherwise. The probability of detection (POD) and probability of false
+        detection (POFD) are calculated for the converted forecast. From the
+        POD and POFD data, the area under the ROC curve is calculated.
+
+        Ideally concave ROC curves should be generated rather than traditional
+        ROC curves.
+
+    Raises:
+        ValueError: if `fcst` contains values outside of the range [0, 1]
+        ValueError: if `obs` contains non-nan values not in the set {0, 1}
+        ValueError: if 'threshold' is a dimension in `fcst`.
+        ValueError: if values in `thresholds` are not montonic increasing or are outside
+          the range [0, 1]
+    """
+    if check_args:
+        if fcst.max().item() > 1 or fcst.min().item() < 0:
+            raise ValueError("`fcst` contains values outside of the range [0, 1]")
+
+        if np.max(thresholds) > 1 or np.min(thresholds) < 0:
+            raise ValueError("`thresholds` contains values outside of the range [0, 1]")
+
+        if not np.all(np.array(thresholds)[1:] >= np.array(thresholds)[:-1]):
+            raise ValueError("`thresholds` is not monotonic increasing between 0 and 1")
+
+    # make a discrete forecast for each threshold in thresholds
+    # discrete_fcst has an extra dimension 'threshold'
+    discrete_fcst = binary_discretise(fcst, thresholds, ">=")
+
+    all_dims = set(fcst.dims).union(set(obs.dims))
+    final_reduce_dims = gather_dimensions(fcst.dims, obs.dims, reduce_dims, preserve_dims)
+    final_preserve_dims = all_dims - set(final_reduce_dims)
+    auc_dims = () if final_preserve_dims is None else tuple(final_preserve_dims)
+    final_preserve_dims = auc_dims + ("threshold",)
+
+    pod = probability_of_detection(
+        discrete_fcst, obs, preserve_dims=final_preserve_dims, weights=weights, check_args=check_args
+    )
+
+    pofd = probability_of_false_detection(
+        discrete_fcst, obs, preserve_dims=final_preserve_dims, weights=weights, check_args=check_args
+    )
+
+    # Need to ensure ordering of dims is consistent for xr.apply_ufunc
+    pod = pod.transpose(*final_preserve_dims)
+    pofd = pofd.transpose(*final_preserve_dims)
+
+    auc = -1 * xr.apply_ufunc(
+        np.trapz,
+        pod,
+        pofd,
+        input_core_dims=[pod.dims, pofd.dims],
+        output_core_dims=[auc_dims],
+        dask="parallelized",
```

Comment:
I think you'll need to add `dask` as one of the [required dependencies here](https://github.com/nci/scores/blob/develop/pyproject.toml#L22)

Reply:
No, it will only use dask if the input data is a dask array. See https://docs.xarray.dev/en/stable/generated/xarray.apply_ufunc.html

I've checked that this runs fine when dask is not in the env and we pass in a standard (no-dask) xarray object

Reply:
Ok, I thought you were also importing dask but I was mistaken.